### PR TITLE
Jmpi 443 map json to fhir

### PR DESCRIPTION
**Pull request by IlyesBenAmara**
- this PR adds an endpoint that can receive a patient record and returns a FHIR Patient resource
to test this feature, send this `PatientRecord` to `http://localhost:50000/JeMPI/fhir`via POST request,
```json
{
    "patientId": "123DDD",
    "demographicData": {
        "auxId": "1234",
        "givenName": "John",
        "familyName": "Doe",
        "gender": "Male",
        "dob": "2023-04-04",
        "city": "Maseru",
        "phoneNumber": "555-223-098",
        "nationalId": "123456yeir"
    },
    "sourceId": {
        "uid": "123",
        "facility": "Some Facility",
        "patient": "John"
    }
}
```
 it should return the following FHIR resource: 

```json
{
  "resourceType": "Patient",
  "identifier": [ {
    "value": "123DDD"
  } ],
  "name": [ {
    "given": [ "John" ]
  }, {
    "family": "Doe"
  } ],
  "birthDate": "2023-04-04",
  "address": [ {
    "city": "Cape town"
  } ]
}
```

### CODE REVIEW

**Comment by ajayyuvo** [comment]
Hi,

Getting this screen after running through the installation instructions for Jembi/JeMPI and jembi/jempi-web.
All containers start successfully.
Last step that I run is ./deploy-platform-local.sh up local-frontend in the jembi/jempi-web project and the screenshot below is what I see. 
Not sure what should we be seeing after starting the front-end application.

Thank you,
Ajay Joshi 

> _ No description provided. _
image

File: JeMPI_Apps/JeMPI_API/pom.xml [hunk 1]
```diff
@@ -195,6 +195,22 @@
...
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
+            <version>2.10.1</version>
+        </dependency>
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+            <version>20211205</version>
+        </dependency>
```
Comment (marrouchi) [inline JeMPI_Apps/JeMPI_API/pom.xml:212]: Do we need both ?

File: JeMPI_Apps/JeMPI_API/src/main/java/org/jembi/jempi/api/BackEnd.java [hunk 2]
```diff
@@ -109,9 +110,26 @@ public Receive<Event> actor() {
...
+   private Behavior<Event> mapToFhir(final MapToFhirRequest request) {
+      PatientRecord payload = request.patientRecord();
+      LOGGER.debug(payload);
+      Patient patient;
+      String patientJson =  "";
+      try {
+         patient = JsonToFhir.mapToPatientFhir(payload);
+         FhirContext ctx = FhirContext.forR4();
+      // Serialize the patient object to FHIR JSON
+         patientJson = ctx.newJsonParser().setPrettyPrint(true).encodeResourceToString(patient);
```
Comment (marrouchi) [inline JeMPI_Apps/JeMPI_API/src/main/java/org/jembi/jempi/api/BackEnd.java:126]: Let's move them inside mapToPatientFhir() function

File: JeMPI_Apps/JeMPI_ETL/pom.xml [hunk 3]
```diff
@@ -118,6 +119,33 @@
...
+            <groupId>ca.uhn.hapi.fhir</groupId>
+            <artifactId>org.hl7.fhir.utilities</artifactId>
+            <version>${fhir_core_version}</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>commons-net</groupId>
+                    <artifactId>commons-net</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
```
Comment (marrouchi) [inline JeMPI_Apps/JeMPI_ETL/pom.xml:137]: fhir imported twice ? don't we need those deps only in the shared pakage ?

File: JeMPI_Apps/JeMPI_ETL/pom.xml [hunk 4]
```diff
@@ -118,6 +119,33 @@
...
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
+            <version>2.10.1</version>
+        </dependency>
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+            <version>20211205</version>
+        </dependency>
```
Comment (marrouchi) [inline JeMPI_Apps/JeMPI_ETL/pom.xml:148]: Do we need those here ?

File: JeMPI_Apps/JeMPI_ETL/src/main/java/org/jembi/jempi/etl/JsonToFhir.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
+package org.jembi.jempi.etl;
+
+import org.hl7.fhir.r4.model.*;
+import org.jembi.jempi.shared.models.CustomDemographicData;
+import org.jembi.jempi.shared.models.PatientRecord;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import java.lang.reflect.Field;
+import static org.apache.commons.text.WordUtils.capitalizeFully;
```
Comment (marrouchi) [inline JeMPI_Apps/JeMPI_ETL/src/main/java/org/jembi/jempi/etl/JsonToFhir.java:9]: Let's not install a package for this, (remove)

File: JeMPI_Apps/JeMPI_ETL/src/main/java/org/jembi/jempi/etl/JsonToFhir.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+public final class JsonToFhir {
+    private static String getFhirPath(final String fieldName, final JSONObject config) {
+        JSONArray fields = config.getJSONArray("fields");
+        for (int i = 0; i < fields.length(); i++) {
+            JSONObject field = fields.getJSONObject(i);
+            if (fieldName.equalsIgnoreCase(capitalizeFully(field.getString("fieldName")).replace("_", ""))) {
+                return field.getString("fhirPath");
```
Comment (marrouchi) [inline JeMPI_Apps/JeMPI_ETL/src/main/java/org/jembi/jempi/etl/JsonToFhir.java:17]: Let's optimize this ..

File: JeMPI_Apps/JeMPI_Shared/pom.xml [hunk 7]
```diff
@@ -71,6 +71,30 @@
...
+        <dependency>
+            <groupId>com.googlecode.json-simple</groupId>
+            <artifactId>json-simple</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+            <version>20211205</version>
+        </dependency>
```
Comment (marrouchi) [inline JeMPI_Apps/JeMPI_Shared/pom.xml:94]: Do we need all three ?

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: JeMPI_Apps/JeMPI_Shared/src/main/java/org/jembi/jempi/shared/mapper/JsonToFhir.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+                        String demoFieldName = demoField.getName();
+                        fieldValue = (String) demoField.get(demographicData);
+                        if (fieldValue != null) {
+                            String fhirPath = getFhirPath(demoFieldName);
+                            if (fhirPath != null) {
+                                processField(patient, fieldValue, fhirPath);
+                            }
+                        }
+                    }
+                } else if (fieldName.equals("patientId")) {
```
Comment (marrouchi) [inline JeMPI_Apps/JeMPI_Shared/src/main/java/org/jembi/jempi/shared/mapper/JsonToFhir.java:89]: Aren't we treating this case in processField() ?

**Review by marrouchi — APPROVED** [review]

**Comment by marrouchi** [comment]
@IlyesBenAmara  Looks good to me as an initial implementation for the fhir mapper. Note that we will need to add other fhir fields,  maybe we could achieve that as we go.

**Comment by MatthewErispe** [comment]
Closing pull request. Duplicated by another

**Pull request closed without merging**